core/server/worker/src/main/java/alluxio/worker/page/BlockPageId.java

```diff
@@ -0,0 +1,191 @@
+   * @param blockId the block ID
+   * @param pageIndex index of the page in the block
+   * @param blockSize block size
+   */
+  public BlockPageId(long blockId, long pageIndex, long blockSize) {
+    super(fileIdOf(blockId, blockSize), pageIndex);
+    mBlockId = blockId;
+    mBlockSize = blockSize;
+  }
+
+  /**
+   * Creates a new page of the temporary block.
+   * @param blockId
+   * @param pageIndex
+   * @return page ID
+   */
+  public static BlockPageId newTempPage(long blockId, long pageIndex) {
+    return new BlockPageId(blockId, pageIndex, INVALID_BLOCK_SIZE);
+  }
+
+  /**
+   * @param blockId
+   * @param blockSize
+   * @return file ID
+   */
+  public static String fileIdOf(long blockId, long blockSize) {
+    return String.format(FILE_ID_TEMPLATE, blockId, blockSize).intern();
+  }
+
+  /**
+   * @param blockId
+   * @return file ID
+   */
+  public static String tempFileIdOf(long blockId) {
+    return fileIdOf(blockId, INVALID_BLOCK_SIZE);
+  }
+
+  /**
+   * @param fileId
+   * @return block ID
+   */
+  public static Optional<Long> parseBlockId(String fileId) {
+    Matcher matcher = FILE_ID_PATTERN.matcher(fileId);
+    if (matcher.matches()) {
+      return Optional.of(Long.parseLong(matcher.group(1)));
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * @param fileId
+   * @return block size
+   */
+  public static Optional<Long> parseBlockSize(String fileId) {
+    Matcher matcher = FILE_ID_PATTERN.matcher(fileId);
+    if (matcher.matches()) {
+      return Optional.of(Long.parseLong(matcher.group(2)));
+    }
+    return Optional.empty();
+  }
+
+  /**
+   * @return the block ID
+   */
+  public long getBlockId() {
+    return mBlockId;
+  }
+
+  /**
+   * @return the block size
+   */
+  public long getBlockSize() {
+    return mBlockSize;
+  }
+
+  /**
+   * Try to downcast a {@link PageId} to a {@link BlockPageId}. If the object is already a block
+   * page ID, then it is immediately returned. Otherwise, attempt to parse the {@code fileId} of
+   * the page if the encoded file name matches a block and contains all necessary metadata.
+   *
+   * @param pageId the page ID to downcast
+   * @return the downcast block page ID
+   * @throws IllegalArgumentException if the page ID cannot be cast to a block page ID
+   */
+  public static BlockPageId tryDowncast(PageId pageId) throws IllegalArgumentException {
+    if (pageId instanceof BlockPageId) {
+      return (BlockPageId) pageId;
+    }
+    String fileId = pageId.getFileId();
+    Matcher match = FILE_ID_PATTERN.matcher(fileId);
+    if (match.matches()) {
+      String blockIdHex = match.group(1);
+      String blockSizeHex = match.group(2);
+      try {
+        long blockId = Long.parseLong(blockIdHex, 16);
+        long blockSize = Long.parseLong(blockSizeHex, 16);
+        return new BlockPageId(blockId, pageId.getPageIndex(), blockSize);
+      } catch (NumberFormatException e) {
+        throw new IllegalArgumentException(
```

Comment:
It's a bit weird here since NumberFormatException is extending IllegalArgumentException. And I'm not a big fan of adding RuntimeException in method signature, java doc should be enough.

Reply:
And I'm also a little confused about expectation here. tryDowncast means we are expecting failure, which seems like Optional is a better option. But doesn't seems like we explicitly catch exception and handle it. So maybe we should just name it as downcast.

Reply:
> tryDowncast means we are expecting failure, which seems like Optional is a better option.

Yeah the intention was to signal to the caller that downcasting failure is expected, rather than an "exception". Using an `Optional`, however, does not give you a detailed error message of what went wrong, what's the part that does not match a paged block ID.

But now since I came to rethink about it, all this code is in the worker's page store, so there should not be any page that does not follow the name pattern. Anything that does is really an exception.

Reply:
fixed in [3206ba3](https://github.com/Alluxio/alluxio/pull/16254/commits/3206ba3bb61e5dfee62a9f244827552eb7e51d70)